lib/src/iterable_matchers.dart

```diff
@@ -206,6 +215,26 @@ class _UnorderedMatches extends Matcher {
   Description describeMismatch(item, Description mismatchDescription,
           Map matchState, bool verbose) =>
       mismatchDescription.add(_test(item));
+
+  /// Returns [true] if the value at [valueIndex] can be paired with some
+  /// unmatched matcher.
+  ///
+  /// Recursively looks for new pairings whenever there is a conflict. [seen]
+  /// tracks the matchers that have already been consumed within this search.
+  bool _findPairing(List<List<bool>> adjacency, int valueIndex, Set<int> seen,
+      List<int> matched) {
+    for (int i = 0; i < matched.length; i++) {
+      if (adjacency[valueIndex][i] && !seen.contains(i)) {
+        seen.add(i);
+        if (matched[i] < 0 ||
+            _findPairing(adjacency, matched[i], seen, matched)) {
```

Comment:
This is really hard to follow, please add a comment describing the logic or simplify it if possible.

```
If we didn't have a previous match, use this as the match and return.

Otherwise < describe logic for recursive call to _findPairing>.
```
Specifically understanding interactions of `seen` and `matched` through repeated recursive calls here is really difficult to understand.

Reply:
Updated the doc comment, added some intermediate variables with descriptive names, and added a block comment at the point where we recurse.  WDUT?